diff --git a/target/linux/ramips/dts/mt7621_xiaomi_mi-router-3g.dts b/target/linux/ramips/dts/mt7621_xiaomi_mi-router-3g.dts
index 40ea6625d..96e834d38 100644
--- a/target/linux/ramips/dts/mt7621_xiaomi_mi-router-3g.dts
+++ b/target/linux/ramips/dts/mt7621_xiaomi_mi-router-3g.dts
@@ -9,7 +9,7 @@
 		led-failsafe = &led_status_red;
 		led-running = &led_status_blue;
 		led-upgrade = &led_status_yellow;
-		label-mac-device = &gmac0;
+		label-mac-device = &ethernet;
 	};
 
 	leds {
@@ -85,28 +85,19 @@
 	};
 };
 
-&gmac0 {
-	mtd-mac-address = <&factory 0xe006>;
+&ethernet {
+	compatible = "mediatek,ralink-mt7621-eth";
+	mediatek,switch = <&gsw>;
+	mtd-mac-address = <&factory 0xe000>;
 };
 
 &switch0 {
-	ports {
-		port@1 {
-			status = "okay";
-			label = "wan";
-			mtd-mac-address = <&factory 0xe000>;
-		};
-
-		port@2 {
-			status = "okay";
-			label = "lan2";
-		};
+	/delete-property/ compatible;
+	phy-mode = "rgmii";
+};
 
-		port@3 {
-			status = "okay";
-			label = "lan1";
-		};
-	};
+&gsw {
+	compatible = "mediatek,ralink-mt7621-gsw";
 };
 
 &state_default {
